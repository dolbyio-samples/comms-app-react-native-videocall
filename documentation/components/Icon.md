# Icon

The Icon component is responsible for displaying icon.

## Props

| Name        | Type                                      | Default | Description                            |
| ----------- | ----------------------------------------- | ------- | -------------------------------------- |
| `name`      | IconComponentName                         | -       | The text to display for the component. |
| `color`     | ColorKey                                  | -       | Tint color for the Icon                |
| `colorTone` | ColorTone ('light' / 'default' / 'dark')  | default | The color tone of the icon.            |
| `size`?     | IconSize ('xxs' / 'xs' / 's' / 'm' / 'l') | m       | The size of the icon.                  |
| `testID`?   | string                                    | -       | The unique E2E test handler.           |

## Available icons

| Name                   | Icon                                                                                                  |
| ---------------------- | ----------------------------------------------------------------------------------------------------- |
| arrowLeft              | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/arrow-left.svg)               |
| cameraOff              | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/camera-off.svg)               |
| cameraReverse          | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/camera-reverse.svg)           |
| camera                 | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/camera.svg)                   |
| chat                   | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/chat.svg)                     |
| close                  | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/close.svg)                    |
| copy                   | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/copy.svg)                     |
| dotsHorizontal         | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/dots-horizontal.svg)          |
| dotsVertical           | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/dots-vertical.svg)            |
| handset                | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/handset.svg)                  |
| headphones             | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/headphones.svg)               |
| info                   | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/info.svg)                     |
| microphoneOff          | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/microphone-off.svg)           |
| microphone             | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/microphone.svg)               |
| participants           | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/participants.svg)             |
| pin                    | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/pin.svg)                      |
| present                | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/present.svg)                  |
| record                 | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/record.svg)                   |
| sendMessage            | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/send-message.svg)             |
| settings               | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/settings.svg)                 |
| speaker                | ![Alt text](../../packages/mobile/src/components/ui/Icon/IconComponents/speaker.svg)                  |

## Examples

### React Native

```javascript
return <Icon name="arrowLeft" size="s" testID="testID" />;
```
