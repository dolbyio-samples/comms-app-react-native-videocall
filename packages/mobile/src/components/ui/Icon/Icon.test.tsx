import { defaultTheme as theme } from '@dolbyio/comms-uikit-common';
import Color from 'color';
import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';

import { render } from '../../../utils/test-utils';

import Icon, { IconSize } from './Icon';

const name = 'settings';
const testID = 'testID';
const size = 's';
const color = 'primary.500';

const IconSizes: { [key in IconSize]: { width: number; height: number } } = {
  xxs: { width: 10, height: 10 },
  xs: { width: 16, height: 16 },
  s: { width: 18, height: 18 },
  m: { width: 24, height: 24 },
  l: { width: 172, height: 172 },
};

const { colors } = theme;

describe('Icon component', () => {
  test('Passes TestID', () => {
    const { getByTestId } = render(<Icon name={name} testID={testID} />);
    expect(getByTestId(testID)).not.toBeNull();
  });
  test('Passes styles for default colortone', () => {
    const { getByTestId } = render(<Icon name={name} testID={testID} color={color} size={size} />);
    const element = getByTestId(testID);
    expect(element).toHaveProp('fill', colors.primary[500]);
    expect(element).toHaveProp('width', IconSizes[size].width);
    expect(element).toHaveProp('height', IconSizes[size].height);
  });
  test('Passes styles for light colortone', () => {
    const { getByTestId } = render(<Icon name={name} colorTone="light" testID={testID} color={color} size={size} />);
    const element = getByTestId(testID);
    expect(element).toHaveProp('fill', Color(colors.primary[500]).lighten(0.2).hex());
    expect(element).toHaveProp('width', IconSizes[size].width);
    expect(element).toHaveProp('height', IconSizes[size].height);
  });
  test('Passes styles for dark colortone', () => {
    const { getByTestId } = render(<Icon name={name} colorTone="dark" testID={testID} color={color} size={size} />);
    const element = getByTestId(testID);
    expect(element).toHaveProp('fill', Color(colors.primary[500]).darken(0.5).hex());
    expect(element).toHaveProp('width', IconSizes[size].width);
    expect(element).toHaveProp('height', IconSizes[size].height);
  });
  test('Uses default size as m(medium)', () => {
    const size = 'm';
    const { getByTestId } = render(<Icon name={name} testID={testID} color={color} />);
    const element = getByTestId(testID);
    expect(element).toHaveProp('fill', colors.primary[500]);
    expect(element).toHaveProp('width', IconSizes[size].width);
    expect(element).toHaveProp('height', IconSizes[size].height);
  });
  test('Supports xxs size', () => {
    const size = 'xxs';
    const { getByTestId } = render(<Icon name={name} testID={testID} color={color} size={size} />);
    const element = getByTestId(testID);
    expect(element).toHaveProp('fill', colors.primary[500]);
    expect(element).toHaveProp('width', IconSizes[size].width);
    expect(element).toHaveProp('height', IconSizes[size].height);
  });
  test('Supports xs size', () => {
    const size = 'xs';
    const { getByTestId } = render(<Icon name={name} testID={testID} color={color} size={size} />);
    const element = getByTestId(testID);
    expect(element).toHaveProp('fill', colors.primary[500]);
    expect(element).toHaveProp('width', IconSizes[size].width);
    expect(element).toHaveProp('height', IconSizes[size].height);
  });
  test('Supports l size', () => {
    const size = 'l';
    const { getByTestId } = render(<Icon name={name} testID={testID} color={color} size={size} />);
    const element = getByTestId(testID);
    expect(element).toHaveProp('fill', colors.primary[500]);
    expect(element).toHaveProp('width', IconSizes[size].width);
    expect(element).toHaveProp('height', IconSizes[size].height);
  });
});

describe('Icon component snapshot', () => {
  test('matches for dark colortone', () => {
    const tree = renderer
      .create(<Icon name={name} colorTone="dark" testID={testID} color={color} size={size} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
