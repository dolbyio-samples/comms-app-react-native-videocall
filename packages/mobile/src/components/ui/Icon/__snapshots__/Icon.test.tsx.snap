// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component snapshot matches for dark colortone 1`] = `
<svg
  data-file-name="SvgSettings"
  fill="#441466"
  height={18}
  testID="testID"
  width={18}
/>
`;
